### PR TITLE
Pass --time to nerdctl stop to avoid timeouts when stopping stuck con…

### PR DESCRIPTION
**Pull request by OneBlue**
…tainers

## Summary of the Pull Request

This change solves multiple issues in the Stop() implementation:

- Use WaitAndCaptureOutput() to consume the nerdctl stop process output to avoid potential pipe deadlocks
- Pass the timeout to nerdctl via --time so that the container is killed after the user provided timeout (this allows a user to pass 0 to immediately kill the container) 

## PR Checklist

- [ ] **Closes:** Link to issue #xxx
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated if needed and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated if needed
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/wsl/) and link it here: #xxx

### CODE REVIEW

File: src/windows/wslaservice/exe/WSLAContainer.cpp [hunk 1]
```diff
@@ -125,12 +125,13 @@ try
         "Container '%hs' is not in a stoppable state: %i",
         m_name.c_str(),
         m_state);
-    ServiceProcessLauncher launcher(nerdctlPath, {nerdctlPath, "stop", m_name});
+    ServiceProcessLauncher launcher(
+        nerdctlPath, {nerdctlPath, "stop", m_name, "--time", std::to_string(static_cast<ULONG>(std::round(TimeoutMs / 1000)))});
```
[the remark this answers is not in the document]
Comment (ptrivedi) [reply src/windows/wslaservice/exe/WSLAContainer.cpp:129]: One thing to keep in mind is that depending on the nerdctl version, the option name for this could be different. For example, here --stop-timeout and --stop-signal are the documented options: https://github.com/containerd/nerdctl/blob/main/docs/command-reference.md#:~:text=%F0%9F%90%B3%20%2D%2Dstop%2Dsignal%3A%20Signal%20to,seconds)%20to%20stop%20a%20container
Nothing to do for now, this is FYI

File: src/windows/wslaservice/exe/WSLAContainer.cpp [hunk 2]
```diff
@@ -125,12 +125,13 @@ try
...
     // TODO: Figure out how we want to handle custom signals and timeout values.
     // nerdctl stop has a --time and a --signal option that can be used
     // By default, it uses SIGTERM and a default timeout of 10 seconds.
-    auto result = launcher.Launch(*m_parentVM).Wait(TimeoutMs);
-    THROW_HR_IF_MSG(E_FAIL, result.first != 0, "%hs", launcher.FormatResult(result.first).c_str());
+    auto result = launcher.Launch(*m_parentVM).WaitAndCaptureOutput();
```
[the remark this answers is not in the document]
Comment (ptrivedi) [reply src/windows/wslaservice/exe/WSLAContainer.cpp:133]: Pass timeout to Wait

File: src/windows/wslaservice/exe/WSLAContainer.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ptrivedi) [reply src/windows/wslaservice/exe/WSLAContainer.cpp:129]: I think this is a good suggestion. Should add test with edge case timeout values

**Review by ptrivedi — APPROVED** [review]
approved for when comments are addressed

File: src/windows/wslaservice/exe/WSLAContainer.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (OneBlue) [reply src/windows/wslaservice/exe/WSLAContainer.cpp:133]: We actually don't want to, because if we hit the SIGKILL timeout, we want to wait until nerdctl has completed that, otherwise we'll get a TIMEOUT error out. 

This happens for instance if the caller passes 0 to immediately kill the container

File: src/windows/wslaservice/exe/WSLAContainer.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (OneBlue) [reply src/windows/wslaservice/exe/WSLAContainer.cpp:129]: Agreed ! I have some in https://github.com/microsoft/WSL/pull/13862, but didn't bring them here to make the fix simpler

File: src/windows/wslaservice/exe/WSLAContainer.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ptrivedi) [reply src/windows/wslaservice/exe/WSLAContainer.cpp:133]: Hmm, looking through all our Wait calls on nerdctl will have infinite timeout. I wonder if our API design should allow timeout on those nerdctl executions

[the remark this answers is not in the document]
Comment (OneBlue) [reply src/windows/wslaservice/exe/WSLAContainer.cpp:133]: Probably in some cases yeah, but I think that should come as a bigger discussion, probably via something that can be configured via the session settings

[the remark this answers is not in the document]
Comment (ptrivedi) [reply src/windows/wslaservice/exe/WSLAContainer.cpp:133]: Yes, let's do a broader discussion. Looks good for now
